src/main/java/org/jenkinsci/plugins/workflow/steps/ErrorInfoStep.java

```diff
@@ -0,0 +1,193 @@
+ */
+public class ErrorInfoStep extends AbstractStepImpl {
+
+    public final Throwable error;
+
+    @DataBoundConstructor public ErrorInfoStep(Throwable error) {
+        this.error = error;
+    }
+
+    public static class Execution extends AbstractSynchronousStepExecution<ErrorInfo> {
+
+        private static final long serialVersionUID = 1;
+        @Inject private transient ErrorInfoStep step;
+        @StepContextParameter private transient FlowExecution execution;
+
+        @Override protected ErrorInfo run() throws Exception {
+            return new ErrorInfo(step.error, execution);
+        }
+
+    }
+
+    @Extension public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+
+        public DescriptorImpl() {
+            super(Execution.class);
+        }
+
+        @Override public String getFunctionName() {
+            return "errorInfo";
+        }
+
+        @Override public String getDisplayName() {
+            return "Calculate information about an error";
+        }
+
+        // TODO blank config.jelly
+
+    }
+
+    public static class ErrorInfo implements Serializable {
+
+        private static final long serialVersionUID = 1;
+        private final Throwable error;
+        private transient FlowExecution execution;
+        private final FlowExecutionOwner executionOwner;
+
+        ErrorInfo(Throwable error, FlowExecution execution) {
+            this.error = error;
+            this.execution = execution;
+            executionOwner = execution.getOwner();
+        }
+
+        private FlowExecution getExecution() throws IOException {
+            if (execution == null) {
+                execution = executionOwner.get();
+            }
+            return execution;
+        }
+
+        /**
+         * Finds a node which threw this exception or one of its causes.
+         * Note that {@link Throwable#equals} is just pointer equality,
+         * which we cannot use since we may be loading deserialized exceptions,
+         * so we compare by stack trace instead.
+         */
+        private @CheckForNull FlowNode getNode() throws IOException {
+            final Set<String> stackTraces = new HashSet<>();
+            for (Throwable t = error; t != null; t = t.getCause()) {
+                stackTraces.add(Functions.printThrowable(t));
+            }
+            Predicate<FlowNode> threwException = new Predicate<FlowNode>() {
+                @Override
+                public boolean apply(FlowNode input) {
+                    if (input instanceof BlockEndNode) {
+                        return false;
+                    }
+                    ErrorAction a = input.getAction(ErrorAction.class);
+                    return (a != null && stackTraces.contains(Functions.printThrowable(a.getError())));
+                }
+            };
+            return new FlowScanner.DepthFirstScanner().findFirstMatch(getExecution(), threwException);
```

Comment:
Once the ForkScanner is complete, it should be far more efficient for this case, because:

1. It doesn't have to keep a HashSet of all visited FlowNodes (just a stack with depth related to the number of levels of parallel branches). 
2. DepthFirstScanner will walk to the start of the flow, if the failure occurs in any branch branch of a parallel block beyond the first.  ForkScanner will only walk the up to the start of the parallel block containing the failure (and then go through the branches there). 

Technically the completely optimal case here is probably to skip blocks without an ErrorAction (if we can guarantee they're attached to all the BlockEndNodes where failures happen) AND to do a proper breadth-first search.